Directives link API documentation did not describe or mention the controller param.

hahla opened the pull request:
The 4th param for link() was not defined in the api doc, but used in the examples. This change adds the param and describes it.

hahla committed:
docs(guide/Directives): describe your change...

The 4th param for link() was not defined in the api doc, but used in the examples. This change adds the param and describes it.

gkalpak commented:
It is not that simple :smile:

Basically, there is a fifth parameter (`transcludeFn`) and the `controller(s)` paremeter value depends on several factors. I suggest you briefly mention the 5 parameters and link to [`$compile` API docs][1] for more info.

@hahla, could you make that change ?

[1]: https://docs.angularjs.org/api/ng/service/$compile#-link-

hahla commented:
Yes sounds good, will update
On Tue, Oct 6, 2015 at 11:43 PM Georgios Kalpakas <notifications@github.com>
wrote:

> It is not that simple [image: :smile:]
>
> Basically, there is a fifth parameter (transcludeFn) and the controller(s)
> paremeter value depends on several factors. I suggest you briefly mention
> the 5 parameters and link to $compile API docs
> <https://docs.angularjs.org/api/ng/service/%24compile#-link-> for more
> info.
>
> @hahla <https://github.com/hahla>, could you make that change ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular.js/pull/13028#issuecomment-146091481>.
>
-- 
Sent from my iPhone

gkalpak commented:
Fantastic !
<sub>Feel free to ping when you're done, so I can take another look.</sub>

Commit:
docs(guide/Directives):

Adds reference to $compile for link function parameters and gives quick intro
to 5th param transcludeFn which was also missing in doc.

hahla commented:
@gkalpak updated and ready for your review

gkalpak commented on docs/content/guide/directive.ngdoc:
```diff
@@ -587,14 +587,15 @@ want to reuse throughout your app.
 * `attrs` is a hash object with key-value pairs of normalized attribute names and their
   corresponding attribute values.
+* `controller` is the controller for the directive, if defined.
+* `transcludeFn` is a transclude linking function pre-bound to the correct translusion scope.
```
_translusion_  -->  _transclusion_

gkalpak commented on docs/content/guide/directive.ngdoc:
```diff
@@ -587,14 +587,15 @@ want to reuse throughout your app.
-`link` takes a function with the following signature, `function link(scope, element, attrs) { ... }`
-where:
+Directives that want to modify the DOM typically use the `link` option, which is documented on the {@link ng.$compile `$compile` API} page.
+`link` takes a function with the following signature, `function link(scope, element, attrs, controller, transcludeFn) { ... }`, where:
```
Please, wrap lines at 100 chars max.

gkalpak commented on docs/content/guide/directive.ngdoc:
```diff
@@ -587,14 +587,15 @@ want to reuse throughout your app.
 * `element` is the jqLite-wrapped element that this directive matches.
 * `attrs` is a hash object with key-value pairs of normalized attribute names and their
   corresponding attribute values.
+* `controller` is the controller for the directive, if defined.
```
This is not 100% accurate. Based on the `$compile` API docs, I would suggest something like this:

`controller` is the directive's required controller instance(s) or it's own controller (if any). The exact value depends on the directive's `require` property:

* no controller(s) required: the directive's own controller, or undefined if it doesn't have one
* string: the controller instance
* array: array of controller instances

gkalpak commented:
@hahla, thx ! I left a couple of comments.
I wonder if can link directly to `https://docs.angularjs.org/api/ng/service/$compile#-link-`.

LGTM with the comments addressed.

Commit:
docs(guide/Directives):

Updates high level overview for directive link function based on PR comments

hahla commented:
@gkalpak few comments:
- I tried to link (no pun intended ;)) to the $compile.link function, but was unable to. AFAICT it is because of the way compile.js is written and link is not a public function.  Would love to link directly to the $compile#methods_link so if you have a suggestion there let me know

- I did another round of edits to improve readability in the context of the article example ("build a directive to display the current time"), while giving the key information on the `link` function. I decided to not copy the full description of the controller parameter because it was going to be a bit too detailed at this point in the docs and would be distracting to have sub-bullets.

-  added an alert-info to the page which gives a clear link for more detail on the $compile API.

<img width="742" alt="screen shot 2015-10-10 at 1 23 43 pm" src="https://cloud.githubusercontent.com/assets/2063061/10413089/9d36894a-6f52-11e5-936d-a009e8329811.png">


Anyway let me know what you think.

Commit:
docs(guide/Directives): link direct to $compile#-link-

hahla commented:
@gkalpak Looks like there was a CI error with the protractor execution, but I don't see any code errors on my side? How do we get Travis CI to re-run the tests, or should I send in a no-op commit to re-trigger the CI test?

gkalpak commented:
LGTM. I restarted the job  (it was indeed a flake).
I'll merge if Travis is happy. 

Thx @hahla for working on this :+1:

gkalpak commented:
Backported to `v1.4.x` as 8a944b0.
